File: android/src/main/kotlin/com/example/dynamic_domain/DynamicDomainPlugin.kt
package com.example.dynamic_domain

import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.Looper
import androidx.core.content.ContextCompat
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import tunnel_core.Tunnel_core
import tunnel_core.LogHandler
import android.os.Handler
import android.os.Looper

/** DynamicDomainPlugin */
class DynamicDomainPlugin :
    FlutterPlugin,
    MethodCallHandler,
    EventChannel.StreamHandler {
    // 用于 Flutter 和 Android 原生之间通信的 MethodChannel
    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "dynamic_domain")
        channel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "dynamic_domain/logs")
        eventChannel.setStreamHandler(this)

        // 设置 Go 的日志处理器
        Tunnel_core.setLogHandler(object : LogHandler {
            override fun onLog(msg: String?) {
                mainHandler.post {
                    eventSink?.success(msg)
                }
            }
        })
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            "init" -> {
                val appId = call.argument<String>("appId")
                // 在实际应用中，我们可能会在这里对 appId 做一些处理
                result.success(null)
            }
            "setEnv" -> {
                val key = call.argument<String>("key")
                val value = call.argument<String>("value")
                Tunnel_core.setEnv(key, value)
                result.success(null)
            }
            "startTunnel" -> {
                // 从参数获取配置
                val config = call.argument<String>("config") ?: "{}"
                
                // 启动前台服务以保持进程存活
                val serviceIntent = Intent(context, TunnelService::class.java)
                ContextCompat.startForegroundService(context, serviceIntent)

                // 注意：网络操作应该在后台线程上进行，但 Go 会启动自己的 goroutines。
                // 但是，gomobile 调用可能会阻塞。
                val status = Tunnel_core.startTunnel(10808, config)
                if (status == "success" || status == "already running") {
                     result.success("127.0.0.1:10808")
                } else {
                     result.error("START_FAILED", status, null)
                }
            }
            "stopTunnel" -> {
                Tunnel_core.stopTunnel()
                // 停止服务
                val serviceIntent = Intent(context, TunnelService::class.java)
                context.stopService(serviceIntent)
                
                result.success(null)
            }
            "setSystemProxy" -> {
                val host = call.argument<String>("host")
                val port = call.argument<Int>("port")
                
                if (host != null && port != null) {
                    System.setProperty("http.proxyHost", host)
                    System.setProperty("http.proxyPort", port.toString())
                    System.setProperty("https.proxyHost", host)
                    System.setProperty("https.proxyPort", port.toString())
                } else {
                    System.clearProperty("http.proxyHost")
                    System.clearProperty("http.proxyPort")
                    System.clearProperty("https.proxyHost")
                    System.clearProperty("https.proxyPort")
                }
                result.success(null)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}

File: ios/Classes/DynamicDomainPlugin.swift
import Flutter
import UIKit
import Tunnel_core

// 修复编译错误：Gomobile 在 Swift 中生成的协议名称可能不带 Protocol 后缀
// 如果编译失败，尝试使用 Tunnel_coreLogHandler
#if canImport(Tunnel_core)
// 这里我们尝试通过 typealias 兼容不同版本的生成代码
// typealias LogHandlerProtocol = Tunnel_coreLogHandlerProtocol 
#endif

public class DynamicDomainPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private var logDelegate: LogDelegate?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "dynamic_domain", binaryMessenger: registrar.messenger())
        let instance = DynamicDomainPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)

        let eventChannel = FlutterEventChannel(name: "dynamic_domain/logs", binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(instance)

        // 设置 Go 的日志处理器
        // 我们需要一个单独的类来继承 NSObject 并实现 Tunnel_coreLogHandlerProtocol
        // 因为 Swift 不支持类的多重继承
        let delegate = LogDelegate()
        instance.logDelegate = delegate // 保持强引用
        delegate.plugin = instance
        Tunnel_coreSetLogHandler(delegate)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
        case "init":
            // let args = call.arguments as? [String: Any]
            // let appId = args?["appId"] as? String
            // 在实际应用中，我们可能会在这里对 appId 做一些处理
            result(nil)
        case "setEnv":
            let args = call.arguments as? [String: Any]
            let key = args?["key"] as? String
            let value = args?["value"] as? String
            Tunnel_coreSetEnv(key, value)
            result(nil)
        case "startTunnel":
            // 启动 Go 隧道
            let args = call.arguments as? [String: Any]
            guard let config = args?["config"] as? String, !config.isEmpty else {
                result(FlutterError(code: "INVALID_CONFIG", message: "Config string is empty", details: nil))
                return
            }
            
            // 简单校验是否为 JSON
            if !config.trimmingCharacters(in: .whitespacesAndNewlines).hasPrefix("{") {
                result(FlutterError(code: "INVALID_CONFIG", message: "Config is not a valid JSON", details: nil))
                return
            }

            let status = Tunnel_coreStartTunnel(config)
            if status == "success" || status == "already running" {
                result("success")
            } else {
                result(FlutterError(code: "START_FAILED", message: status, details: nil))
            }
        case "stopTunnel":
            Tunnel_coreStopTunnel()
            result(nil)
        case "getStats":
            let stats = Tunnel_coreGetStats()
            result(stats)
        case "validateConfig":
            let args = call.arguments as? [String: Any]
            let config = args?["config"] as? String ?? "{}"
            let status = Tunnel_coreValidateConfig(config)
            result(status)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - FlutterStreamHandler

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
    
    // 内部方法，用于处理来自代理的日志
    fileprivate func sendLog(_ msg: String?) {
        DispatchQueue.main.async {
            self.eventSink?(msg)
        }
    }
}

// 单独的类来处理日志，以避免多重继承问题
// 由于 Swift 中协议与类名冲突，我们尝试通过 extension 或直接声明来解决
// 如果编译失败，说明编译器无法区分 Tunnel_coreLogHandler 类和协议
class LogDelegate: NSObject {
    weak var plugin: DynamicDomainPlugin?
    
    func sendLog(_ msg: String?) {
        plugin?.sendLog(msg)
    }
}

extension LogDelegate: Tunnel_coreLogHandlerProtocol {
    public func onLog(_ msg: String?) {
        self.sendLog(msg)
    }
}
